newhelm/typed_data.py

```diff
@@ -0,0 +1,32 @@
+from typing import Any, Dict, Type, TypeVar
+from typing_extensions import Self
+from pydantic import BaseModel
+
+
+_BaseModelType = TypeVar("_BaseModelType", bound=BaseModel)
+
+
+class TypedData(BaseModel):
+    """This is a generic container that allows Pydantic to do polymorphic serialization.
+
+    This is useful in situations where you have an unknown set of classes that could be
+    used in a particular field.
+    """
+
+    type: str
+    data: Dict[str, Any]
+
+    @classmethod
+    def from_instance(cls, obj: _BaseModelType) -> Self:
```

Comment:
I'm just double-checking this as I'm sure there is a good reason, but why not just `BaseModel`?

```
def from_instance(cls, obj: BaseModel) -> Self
```

Reply:
We need this in `to_instance` in order to get strong typing of the return. I figured symmetry was worth the extra indirection, but I'm happy to update if you think its more clear with them being different.

Reply:
I should have been a bit more clear on this one. Sorry about that.

So in the `to_instance` call, I totally understand the use because we need to type the return value of the `to_instance` method. But in `from_instance` we are not making use of it because we're returning `Self`. I was only specifically referring to the `from_instance` method. It technically works either way, so this is not a blocker for me. Was just considering making it a bit more clear.

Reply:
To clarify, my suggestion was to change `from_instance` but to leave `to_instance` as it is.

Reply:
Done.